### ISSUE TITLE
can't build ole32-sys with nightly-gnu

### ISSUE THREAD

**Issue by Eh2406** [rust-lang/rust#41589]
HI,

Setup:
`Cargo.toml`
```toml
[package]
name = "test"
version = "0.1.0"
authors = ["Eh2406 <YeomanYaacov@gmail.com>"]

[dependencies]
ole32-sys = "0.2.0"
```
`src/main.rs`
```rust
extern crate ole32;
fn main() {
    println!("");
}
```

Expected output:
No errors, as there are no errors when I run `cargo +nightly-msvc build`

Actual output:
```cmd
C:\test>cargo +nightly-gnu build
   Compiling winapi v0.2.8
   Compiling winapi-build v0.1.1
   Compiling ole32-sys v0.2.0
error: linking with `gcc` failed: exit code: 3221225781
  |
  = note: "gcc" "-Wl,--enable-long-section-names" "-fno-use-linker-plugin" "-Wl,--nxcompat" "-nostdlib" "-m64" "C:\\Users\\finkelman.SEMCOGDOM\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\crt2.o" "C:\\Users\\finkelman.SEMCOGDOM\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsbegin.o" "-L" "C:\\Users\\finkelman.SEMCOGDOM\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "C:\\test\\target\\debug\\build\\ole32-sys-6a9f3e9c54364a93\\build_script_build-6a9f3e9c54364a93.0.o" "-o" "C:\\test\\target\\debug\\build\\ole32-sys-6a9f3e9c54364a93\\build_script_build-6a9f3e9c54364a93.exe" "-Wl,--gc-sections" "-nodefaultlibs" "-L" "C:\\test\\target\\debug\\deps" "-L" "C:\\Users\\finkelman.SEMCOGDOM\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib" "-Wl,-Bstatic" "C:\\test\\target\\debug\\deps\\libbuild-24835e0f0d5a822a.rlib" "C:\\Users\\finkelman.SEMCOGDOM\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libstd-e2451a256f66ac19.rlib" "C:\\Users\\finkelman.SEMCOGDOM\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\librand-c279a51d66700350.rlib" "C:\\Users\\finkelman.SEMCOGDOM\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libcollections-d7bf31a4ca1ea637.rlib" "C:\\Users\\finkelman.SEMCOGDOM\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libstd_unicode-d367c3ba0db49600.rlib" "C:\\Users\\finkelman.SEMCOGDOM\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libpanic_unwind-2d4bf02140c11dcb.rlib" "C:\\Users\\finkelman.SEMCOGDOM\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libunwind-add7a84d7e82d084.rlib" "C:\\Users\\finkelman.SEMCOGDOM\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\liblibc-84688accbc86d6b7.rlib" "C:\\Users\\finkelman.SEMCOGDOM\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\liballoc-fe2e68b21f0bdd7a.rlib" "C:\\Users\\finkelman.SEMCOGDOM\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\liballoc_system-7fc0381594c93f56.rlib" "C:\\Users\\finkelman.SEMCOGDOM\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libcore-ea9d77e7c23fe65c.rlib" "C:\\Users\\finkelman.SEMCOGDOM\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\libcompiler_builtins-91b619d34dd1f5aa.rlib" "-Wl,-Bdynamic" "-l" "advapi32" "-l" "ws2_32" "-l" "userenv" "-l" "shell32" "-Wl,-Bstatic" "-l" "gcc_eh" "-l" "pthread" "-Wl,-Bdynamic" "-lmingwex" "-lmingw32" "-lgcc" "-lmsvcrt" "-luser32" "-lkernel32" "C:\\Users\\finkelman.SEMCOGDOM\\.rustup\\toolchains\\nightly-x86_64-pc-windows-gnu\\lib\\rustlib\\x86_64-pc-windows-gnu\\lib\\rsend.o"
  = note:

error: aborting due to previous error

error: Could not compile `ole32-sys`.
Build failed, waiting for other jobs to finish...
error: build failed
```

1.18.0-nightly (036983201 2017-04-26), on windows.

Note: that it works with:
- `cargo +nightly-2017-04-20 build`
- `cargo +nightly-2017-04-24 build`
- `cargo +nightly-2017-04-25 build`

does not work with:
- `cargo +nightly-2017-04-26 build`

**Comment by TimNN** [comment]
I gonna take a wild guess and say that this is likely related to the LLVM 4.0 upgrade (#40123) or more specifically the switch from MinGW-with-win32-threads to MinGW-with-posix-threads.

@Eh2406: Could tell us the the MinGW / `gcc --version` you are using?

**Comment by ollie27** [comment]
Looks like this is the same as https://github.com/rust-lang/rust/issues/31840#issuecomment-297514718 and so should be fixed by #41571.

**Comment by Eh2406** [comment]
Works with the new nightlies, so guess you got it perfect! Thanks!
